File: frontend/src/api/promptHub.ts
const API_URL = (
  import.meta.env.VITE_PROMPTHUB_API_URL ?? "http://127.0.0.1:8011"
).replace(/\/$/, "");

export type PromptHubVisibility = "private" | "unlisted" | "public";
export type PromptHubSort = "latest" | "trending" | "top";

export type PromptHubSharedScope = {
  include_prompt: boolean;
  include_response: boolean;
  include_files: boolean;
  include_diff: boolean;
  include_terminal: boolean;
  include_project_context: boolean;
};

export type PromptHubMetrics = {
  events_count?: number;
  files_changed?: number;
  lines_added?: number;
  lines_removed?: number;
  [key: string]: number | undefined;
};

export type PromptHubListItem = {
  category: string | null;
  id: string;
  metrics: PromptHubMetrics;
  model_name: string | null;
  published_at: string | null;
  score_overall: number | null;
  slug: string;
  summary: string | null;
  tags: string[];
  title: string;
  tool_name: string | null;
};

export type PromptHubFile = {
  additions: number;
  change_type: string | null;
  deletions: number;
  diff: string | null;
  file_path: string;
  id: string;
  is_included: boolean;
  language: string | null;
};

export type PromptHubDetail = PromptHubListItem & {
  comments_count: number;
  created_at: string;
  files: PromptHubFile[];
  prompt_text: string;
  reactions_count: number;
  result_summary: string | null;
  score_architecture: number | null;
  score_backend: number | null;
  score_documentation: number | null;
  score_frontend: number | null;
  score_refactoring: number | null;
  shared_scope: PromptHubSharedScope;
  status: "draft" | "published" | "archived";
  updated_at: string;
  visibility: PromptHubVisibility;
};

export type PromptHubListParams = {
  category?: string;
  limit?: number;
  model?: string;
  offset?: number;
  q?: string;
  sort?: PromptHubSort;
  tag?: string;
};

export type CreatePromptHubDraftRequest = {
  activity_id: string;
  include_diff: boolean;
  include_files: boolean;
  include_project_context: boolean;
  include_prompt: boolean;
  include_response: boolean;
  include_terminal: boolean;
  project_id: string;
  summary?: string | null;
  title: string;
};

export type UpdatePromptHubRequest = {
  category?: string | null;
  prompt_text?: string;
  result_summary?: string | null;
  score_architecture?: number | null;
  score_backend?: number | null;
  score_documentation?: number | null;
  score_frontend?: number | null;
  score_overall?: number | null;
  score_refactoring?: number | null;
  shared_scope?: PromptHubSharedScope;
  summary?: string | null;
  tags?: string[];
  title?: string;
  visibility?: PromptHubVisibility;
};

type ApiValidationItem = {
  loc?: Array<string | number>;
  msg?: string;
  type?: string;
};

type ApiErrorPayload = {
  detail?: string | ApiValidationItem[] | { msg?: string };
};

export class PromptHubApiError extends Error {
  status: number;
  validationErrors: string[];

  constructor(
    message: string,
    {
      status,
      validationErrors = [],
    }: {
      status: number;
      validationErrors?: string[];
    },
  ) {
    super(message);
    this.name = "PromptHubApiError";
    this.status = status;
    this.validationErrors = validationErrors;
  }
}

function parseErrorPayload(payload: ApiErrorPayload | null, status: number) {
  if (typeof payload?.detail === "string") {
    return {
      message: payload.detail,
      validationErrors: [],
    };
  }

  if (Array.isArray(payload?.detail)) {
    const validationErrors = payload.detail
      .map((item) => {
        const path = item.loc?.slice(1).join(".");
        return path && item.msg ? `${path}: ${item.msg}` : item.msg;
      })
      .filter((message): message is string => Boolean(message));

    return {
      message:
        validationErrors.length > 0
          ? validationErrors.join("\n")
          : "Validation failed",
      validationErrors,
    };
  }

  if (
    payload?.detail &&
    typeof payload.detail === "object" &&
    typeof payload.detail.msg === "string"
  ) {
    return {
      message: payload.detail.msg,
      validationErrors: [payload.detail.msg],
    };
  }

  return {
    message: `Prompt Hub request failed with HTTP ${status}`,
    validationErrors: [],
  };
}

async function parseJsonResponse<T>(response: Response): Promise<T> {
  const text = await response.text();
  const payload = text
    ? (() => {
        try {
          return JSON.parse(text) as unknown;
        } catch {
          return null;
        }
      })()
    : null;

  if (response.ok) {
    return payload as T;
  }

  const error = parseErrorPayload(payload as ApiErrorPayload | null, response.status);
  throw new PromptHubApiError(error.message, {
    status: response.status,
    validationErrors: error.validationErrors,
  });
}

export function promptHubErrorMessage(error: unknown, fallback: string) {
  if (error instanceof PromptHubApiError) {
    return error.message || fallback;
  }

  return error instanceof Error ? error.message : fallback;
}

export async function listPublishedPrompts(
  params: PromptHubListParams = {},
  signal?: AbortSignal,
) {
  const search = new URLSearchParams();
  if (params.category && params.category !== "All") {
    search.set("category", params.category);
  }
  if (params.limit) {
    search.set("limit", String(params.limit));
  }
  if (params.model) {
    search.set("model", params.model);
  }
  if (params.offset) {
    search.set("offset", String(params.offset));
  }
  if (params.q?.trim()) {
    search.set("q", params.q.trim());
  }
  if (params.sort) {
    search.set("sort", params.sort);
  }
  if (params.tag) {
    search.set("tag", params.tag);
  }

  const suffix = search.toString() ? `?${search}` : "";
  const response = await fetch(`${API_URL}/api/prompt-hub${suffix}`, {
    credentials: "include",
    signal,
  });
  return parseJsonResponse<PromptHubListItem[]>(response);
}

export async function getPublishedPrompt(slug: string, signal?: AbortSignal) {
  const response = await fetch(
    `${API_URL}/api/prompt-hub/${encodeURIComponent(slug)}`,
    {
      credentials: "include",
      signal,
    },
  );
  return parseJsonResponse<PromptHubDetail>(response);
}

export async function createPromptDraftFromActivity(
  payload: CreatePromptHubDraftRequest,
) {
  const response = await fetch(`${API_URL}/api/prompt-hub/drafts/from-activity`, {
    body: JSON.stringify(payload),
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    method: "POST",
  });
  return parseJsonResponse<PromptHubDetail>(response);
}

export async function updatePublishedPrompt(
  id: string,
  payload: UpdatePromptHubRequest,
) {
  const response = await fetch(`${API_URL}/api/prompt-hub/${id}`, {
    body: JSON.stringify(payload),
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    method: "PATCH",
  });
  return parseJsonResponse<PromptHubDetail>(response);
}

export async function publishPrompt(id: string) {
  const response = await fetch(`${API_URL}/api/prompt-hub/${id}/publish`, {
    credentials: "include",
    method: "POST",
  });
  return parseJsonResponse<PromptHubDetail>(response);
}

export async function archivePrompt(id: string) {
  const response = await fetch(`${API_URL}/api/prompt-hub/${id}/archive`, {
    credentials: "include",
    method: "POST",
  });
  return parseJsonResponse<PromptHubDetail>(response);
}

export const listPromptHubPrompts = listPublishedPrompts;
export const readPromptHubPrompt = getPublishedPrompt;
export const createPromptHubDraft = createPromptDraftFromActivity;
export const updatePromptHubPrompt = updatePublishedPrompt;
export const publishPromptHubPrompt = publishPrompt;

File: backend/app/schemas/prompt_hub.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field, validator

PromptVisibility = Literal["private", "unlisted", "public"]
PromptStatus = Literal["draft", "published", "archived"]
PromptSort = Literal["latest", "trending", "top"]


class PromptHubSharedScope(BaseModel):
    include_prompt: bool = True
    include_response: bool = True
    include_files: bool = False
    include_diff: bool = False
    include_terminal: bool = False
    include_project_context: bool = False


class PromptHubDraftFromActivityRequest(BaseModel):
    project_id: UUID
    activity_id: UUID
    title: str = Field(..., min_length=1, max_length=255)
    summary: str | None = Field(default=None, max_length=4000)
    include_prompt: bool = True
    include_response: bool = True
    include_files: bool = False
    include_diff: bool = False
    include_terminal: bool = False
    include_project_context: bool = False

    @validator("title", "summary", pre=True)
    def strip_string(cls, value: Any) -> Any:
        return value.strip() if isinstance(value, str) else value

    def shared_scope(self) -> PromptHubSharedScope:
        return PromptHubSharedScope(
            include_prompt=self.include_prompt,
            include_response=self.include_response,
            include_files=self.include_files,
            include_diff=self.include_diff,
            include_terminal=self.include_terminal,
            include_project_context=self.include_project_context,
        )


class PromptHubUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    summary: str | None = Field(default=None, max_length=4000)
    prompt_text: str | None = Field(default=None, min_length=1)
    result_summary: str | None = Field(default=None, max_length=12000)
    category: str | None = Field(default=None, max_length=120)
    tags: list[str] | None = Field(default=None, max_items=20)
    visibility: PromptVisibility | None = None
    shared_scope: PromptHubSharedScope | None = None
    score_overall: float | None = Field(default=None, ge=0, le=100)
    score_frontend: float | None = Field(default=None, ge=0, le=100)
    score_backend: float | None = Field(default=None, ge=0, le=100)
    score_architecture: float | None = Field(default=None, ge=0, le=100)
    score_refactoring: float | None = Field(default=None, ge=0, le=100)
    score_documentation: float | None = Field(default=None, ge=0, le=100)

    @validator("title", "summary", "prompt_text", "result_summary", "category", pre=True)
    def strip_optional_string(cls, value: Any) -> Any:
        return value.strip() if isinstance(value, str) else value

    @validator("tags")
    def normalize_tags(cls, value: list[str] | None) -> list[str] | None:
        if value is None:
            return None

        normalized: list[str] = []
        seen: set[str] = set()
        for raw_tag in value:
            tag = raw_tag.strip().lower()
            if not tag or tag in seen:
                continue
            if len(tag) > 40:
                raise ValueError("tags must be 40 characters or fewer")
            seen.add(tag)
            normalized.append(tag)
        return normalized


class PromptHubFileRead(BaseModel):
    id: UUID
    file_path: str
    change_type: str | None
    language: str | None
    diff: str | None
    additions: int
    deletions: int
    is_included: bool

    class Config:
        from_attributes = True


class PromptHubListItem(BaseModel):
    id: UUID
    title: str
    slug: str
    summary: str | None
    model_name: str | None
    tool_name: str | None
    category: str | None
    tags: list[str]
    score_overall: float | None
    metrics: dict[str, Any]
    published_at: datetime | None


class PromptHubDetail(PromptHubListItem):
    prompt_text: str
    result_summary: str | None
    visibility: PromptVisibility
    status: PromptStatus
    shared_scope: dict[str, Any]
    score_frontend: float | None
    score_backend: float | None
    score_architecture: float | None
    score_refactoring: float | None
    score_documentation: float | None
    files: list[PromptHubFileRead]
    comments_count: int
    reactions_count: int
    created_at: datetime
    updated_at: datetime
